core/systemcontractupgrade/system-contract-upgrade.go

```diff
@@ -0,0 +1,116 @@
+			},
+		},
+	}
+}
+
+func UpgradeBuildInSystemContract(config *params.ChainConfig, blockNumber *big.Int, statedb *state.StateDB) {
+	if config == nil || blockNumber == nil || statedb == nil {
+		return
+	}
+	var network string
+	switch core.GenesisHash {
+	/* Add mainnet genesis hash */
+	case params.ChapelGenesisHash:
+		network = chapel
+	case params.RialtoGenesisHash:
+		network = rialto
+	}
+
+	logger := log.New("system-contract-upgrade")
+	if config.IsOnRamanujan(blockNumber) {
+		applySystemContractUpgrade(ramanujanUpgrade[network], blockNumber, statedb, logger)
+	}
+
+	/*
+		apply other upgrades
+	*/
+}
+
+func applySystemContractUpgrade(upgrade *Upgrade, blockNumber *big.Int, statedb *state.StateDB, logger log.Logger) {
+	if upgrade == nil {
+		return
+	}
+
+	logger.Info(fmt.Sprintf("Apply upgrade %s at height %d", upgrade.UpgradeName, blockNumber.Int64()))
+	for _, cfg := range upgrade.Configs {
+		logger.Info(fmt.Sprintf("Upgrade contract %s to commit %s", cfg.ContractAddr.String(), cfg.CommitUrl))
+
+		if cfg.BeforeUpgrade != nil {
+			cfg.BeforeUpgrade(blockNumber, cfg.ContractAddr, statedb)
+		}
+
+		newContractCode, err := hex.DecodeString(cfg.Code)
+		if err != nil {
+			logger.Info(fmt.Sprintf("failed to decode new contract code: %s", err.Error()))
```

Comment:
Let's panic here if there is an error. It is a very serious bug and need stop the node if happened